src/python/pants/backend/python/subsystems/python_native_code.py

```diff
@@ -0,0 +1,125 @@
+  def native_toolchain(self):
+    return NativeToolchain.scoped_instance(self)
+
+  @memoized_property
+  def _python_setup(self):
+    return PythonSetup.scoped_instance(self)
+
+  def pydist_has_native_sources(self, target):
+    return target.has_sources(extension=tuple(self._native_source_extensions))
+
+  def native_target_has_native_sources(self, target):
+    return target.has_sources()
+
+  @memoized_property
+  def _native_target_matchers(self):
+    return {
+      Exactly(PythonDistribution): self.pydist_has_native_sources,
+      Exactly(NativeLibrary): self.native_target_has_native_sources,
+    }
+
+  def _any_targets_have_native_sources(self, targets):
+    # TODO(#5949): convert this to checking if the closure of python requirements has any
+    # platform-specific packages (maybe find the platforms there too?).
+    for tgt in targets:
+      for type_constraint, target_predicate in self._native_target_matchers.items():
+        if type_constraint.satisfied_by(tgt) and target_predicate(tgt):
+          return True
+    return False
+
+  def get_targets_by_declared_platform(self, targets):
+    """
+    Aggregates a dict that maps a platform string to a list of targets that specify the platform.
+    If no targets have platforms arguments, return a dict containing platforms inherited from
+    the PythonSetup object.
+
+    :param tgts: a list of :class:`Target` objects.
+    :returns: a dict mapping a platform string to a list of targets that specify the platform.
+    """
+    targets_by_platforms = defaultdict(list)
+
+    for tgt in targets:
+      for platform in tgt.platforms:
+        targets_by_platforms[platform].append(tgt)
+
+    if not targets_by_platforms:
+      for platform in self._python_setup.platforms:
+        targets_by_platforms[platform] = ['(No target) Platform inherited from either the '
+                                          '--platforms option or a pants.ini file.']
+    return targets_by_platforms
+
+  _PYTHON_PLATFORM_TARGETS_CONSTRAINT = Exactly(PythonBinary, PythonDistribution)
+
+  def check_build_for_current_platform_only(self, targets):
+    """
+    Performs a check of whether the current target closure has native sources and if so, ensures
+    that Pants is only targeting the current platform.
+
+    :param tgts: a list of :class:`Target` objects.
+    :return: a boolean value indicating whether the current target closure has native sources.
+    :raises: :class:`pants.base.exceptions.IncompatiblePlatformsError`
+    """
+    if not self._any_targets_have_native_sources(targets):
+      return False
+
+    targets_with_platforms = filter(self._PYTHON_PLATFORM_TARGETS_CONSTRAINT.satisfied_by, targets)
+    platforms_with_sources = self.get_targets_by_declared_platform(targets_with_platforms)
+    platform_names = platforms_with_sources.keys()
+
+    # There will always be at least 1 platform, because we checked that they have native sources.
+    assert(len(platform_names) >= 1)
```

Comment:
Would be great to add a string here explaining the assertion error.

Reply:
Made it into an exception!